Bluetooth: Mesh: Disable PB-GATT on PB-ADV link

HaavardRei opened the pull request:
Previously, PB-GATT unprovisioned advs would continue being sent if a device is provisioned or in the process of being provisioned (an active provisioning link) over PB-ADV.

PB-GATT advs are now stopped when a PB-ADV provisioning link is established. If the link closes without the device being provisioned, the PB-GATT advs will be re-enabled.

m-alperen-sener commented on subsys/bluetooth/mesh/provisionee.c:
```diff
@@ -657,6 +659,12 @@ static void prov_link_closed(enum prov_bearer_link_status status)
+
+	/* Re-enable PB-GATT advs if they were stopped due to a PB-ADV link. */
+	if (pb_adv_disabled_pb_gatt && !bt_mesh_is_provisioned()) {
+		bt_mesh_prov_enable(BT_MESH_PROV_GATT);
```
besides calling this seems like not safe since it will call [bt_mesh_adv_enable](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/adv_ext.c#L520) through PB-GATT [link_accept](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/pb_gatt.c#L183).

HaavardRei replied:
Hmm I guess we could use just:
(void)bt_mesh_pb_gatt_srv_enable();
bt_mesh_adv_gatt_update();

(Think we might need the update part to start sending?)

HaavardRei replied:
The idea was that bt_mesh_pb_gatt_srv_disable was corresponding to bt_mesh_prov_disable, complementary to bt_mesh_prov_enable (but this would return early due to an existing provisioning link)

m-alperen-sener replied:
> Hmm I guess we could use just: (void)bt_mesh_pb_gatt_srv_enable(); bt_mesh_adv_gatt_update();
> 
> (Think we might need the update part to start sending?)

Sounds right!

> The idea was that bt_mesh_pb_gatt_srv_disable was corresponding to bt_mesh_prov_disable, complementary to bt_mesh_prov_enable (but this would return early due to an existing provisioning link)

bt_mesh_prov enabled/disable might do the job here but it is not intended to do that, at least this is what I assume. Eventually is it not safe to call here anyway.

m-alperen-sener replied:
Sorry for partially delayed comment:

I wonder why bt_mesh_adv_gatt_update() is not called inside bt_mesh_pb_gatt_srv_enable, seems like this is also wrong since it depends the return of bt_mesh_pb_gatt_srv_enable. Maybe we should move bt_mesh_adv_gatt_update to  bt_mesh_pb_gatt_srv_enable as it is done in bt_mesh_pb_gatt_srv_disable.

HaavardRei replied:
In this case I could just use bt_mesh_pb_gatt_srv_disable and bt_mesh_pb_gatt_srv_enable (and enable already checks whether the device has been provisioned). But I don't think we then need a suspend/resume functionality in pb_gatt as we only want to stop the server?

m-alperen-sener replied:
it is you decision, I'd prefer to just represent the actual intention with inline call etc. , which is suspending the PG-GATT until link expires successful or unsuccessful and resume if it is necessary. I am ok with your decision. 👍

HaavardRei commented on subsys/bluetooth/mesh/provisionee.c:
```diff
@@ -657,6 +659,12 @@ static void prov_link_closed(enum prov_bearer_link_status status)
 	bt_mesh_prov_reset_state();
+
+	/* Re-enable PB-GATT advs if they were stopped due to a PB-ADV link. */
+	if (pb_adv_disabled_pb_gatt && !bt_mesh_is_provisioned()) {
```
I think we need to keep track of whether PB-GATT has been disabled through link_open. I. e. if PB-GATT configs are enabled, but you haven't activated PB-GATT advs just calling `bt_mesh_pb_gatt_srv_enable` will turn on the advs which would be unexpected from the user's perspective. (Or if you explicitly turned them off, they will suddenly reappear if a PB-ADV provisioning fails)

m-alperen-sener replied:
my suggestion for that part is to add suspend and resume to PB-GATT (pb_gatt.c), which would be more readable and represent actually what is being actually done. And call call it from here. Thus we can make sure we keep the status of the PB-GATT service state as it is.

alxelax commented on subsys/bluetooth/mesh/provisionee.c:
```diff
@@ -665,6 +673,15 @@ static void prov_link_opened(void)
+	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT) && !is_pb_gatt()) {
+		int err = bt_mesh_pb_gatt_srv_disable();
+
+		if (err != -EALREADY) {
```
```suggestion
		if (!bt_mesh_pb_gatt_srv_disable()) {
```

alxelax reviewed (commented):
In general, this is Ok. However, I kind of agree with Alperen if these are two small static functions like: bt_gatt suspend\resume then overall the source code will be self-explaining.

PavelVPV reviewed (changes requested):
I'm not sure if this solves the opposite problem: if provisioning over PB-GATT has been started and then the provisioning invitation is received over PB-ADV. The code flow will be like this then:
- bt_mesh_pb_adv_recv called [here](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/adv.c#L348)
- pb_adv link is [not active](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L988-L991), thus gen_prov_recv is [called](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L994)
- it calls [gen_prov_start](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L574C13-L574C27)
- `prov.c::prov_recv` is called [here](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L430)
So if I didn't miss anything, similar change is needed for PB-ADV.

This also doesn't consider the case where PB-GATT was disabled through `bt_mesh_prov_disable` or if it was ever enabled before.

Thus, I'd suggest the following:
- introduce `link_cancel` callback in `struct prov_bearer`
- implement `link_cancel` callback in `pb_gatt.c` and `pb_adv.c`:
    - in `pb_gatt.c` in addition to https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/provisionee.c#L742 call `bt_mesh_adv_update()` (it should stop advertising if server is not registered).
    - in `pb_adv.c` do this: https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/provisionee.c#L736-L737
- We can skip this callback for `pb_remote` since noone asked for an ability to disable pb remote after it was enabled, but if we want,  it can do symmetrically to what `node_refresh_link_accept` does, reset the callback and callback data: https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/rpr_srv.c#L1356-L1357
- in `provisionee.c` in `bt_mesh_prov_disable`, call `->link_cancel` for each bearer (PB-GATT, PB-ADV, PB-REMOTE) instead of doing what it currently does
- in `provisionee.c` store bearers enabled through `bt_mesh_prov_enable` and disabled through `bt_mesh_prov_disable`
- in `prov_link_opened` callback in `provisionee.c`, call `->link_cancel` for the bearer that is enabled, but not used for provisioning.
- in `prov_link_closed` callback in `provisisonee.c` restore stopped bearers (or simply call `link_accept` for all bearers, this implies that bearers can understand that they are active and don't need to restart).
- we may not need [this](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/main.c#L640) and [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/settings.c#L108) after the change as if PB-ADV opened the link, PB-GATT link will be stopped in `provisionee.c`.

I could miss something, but I hoped I managed to outline the concept. The idea is to hide logic required to start and stop bearers inside the bearers themselves, and not do this here.

m-alperen-sener commented:
> I'm not sure if this solves the opposite problem: if provisioning over PB-GATT has been started and then the provisioning invitation is received over PB-ADV. The code flow will be like this then:
> 
> * bt_mesh_pb_adv_recv called [here](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/adv.c#L348)
> * pb_adv link is [not active](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L988-L991), thus gen_prov_recv is [called](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L994)
> * it calls [gen_prov_start](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L574C13-L574C27)
> * `prov.c::prov_recv` is called [here](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/pb_adv.c#L430)
>   So if I didn't miss anything, similar change is needed for PB-ADV.

PB-ADV stops advertising in case there is an active link [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/beacon.c#L514), also adding the same check https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/pb_adv.c#L971 prevent any invitation from an PB-ADV provisioner. Thus we can make certain that fisrt bearer to open link is undisturbed by the other.  

> 
> This also doesn't consider the case where PB-GATT was disabled through `bt_mesh_prov_disable` or if it was ever enabled before.
> 
> Thus, I'd suggest the following:
> 
> * introduce `link_cancel` callback in `struct prov_bearer`
> * implement `link_cancel` callback in `pb_gatt.c` and `pb_adv.c`:
>   
>   * in `pb_gatt.c` in addition to https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/provisionee.c#L742
>      call `bt_mesh_adv_update()` (it should stop advertising if server is not registered).
>   * in `pb_adv.c` do this: https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/provisionee.c#L736-L737
> * We can skip this callback for `pb_remote` since noone asked for an ability to disable pb remote after it was enabled, but if we want,  it can do symmetrically to what `node_refresh_link_accept` does, reset the callback and callback data: https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/rpr_srv.c#L1356-L1357
> * in `provisionee.c` in `bt_mesh_prov_disable`, call `->link_cancel` for each bearer (PB-GATT, PB-ADV, PB-REMOTE) instead of doing what it currently does
> * in `provisionee.c` store bearers enabled through `bt_mesh_prov_enable` and disabled through `bt_mesh_prov_disable`
> * in `prov_link_opened` callback in `provisionee.c`, call `->link_cancel` for the bearer that is enabled, but not used for provisioning.
> * in `prov_link_closed` callback in `provisisonee.c` restore stopped bearers (or simply call `link_accept` for all bearers, this implies that bearers can understand that they are active and don't need to restart).
> * we may not need [this](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/main.c#L640) and [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/settings.c#L108) after the change as if PB-ADV opened the link, PB-GATT link will be stopped in `provisionee.c`.
> 
> I could miss something, but I hoped I managed to outline the concept. The idea is to hide logic required to start and stop bearers inside the bearers themselves, and not do this here.

If we can stop a bearer advertising when there is an active link and or creating another link in race condition. then we might not need to handle all the conditions above, individually.

PavelVPV reviewed (commented):
Thanks, @HaavardRei !

Could you please check if we still need to call `bt_mesh_pb_gatt_srv_disable` we may not need [here](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/main.c#L640) and [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/settings.c#L108)?

I think the one called from `bt_mesh_start` is now covered by your change in `prov_link_opened` callback.

I'm not really sure what the one called from `mesh_commit` could does as we would go there only from `settings_load()` when device is provisioned. This (going to `mesh_commit`) is only (unless I miss something) possible if device was just boot. But at this moment the service couldn't be registered yet.

There is a additional case with `bt_mesh_suspend`/`bt_mesh_resume`, I'll create a separate  task for this as they ignore if PB-GATT was ever enabled.

HaavardRei commented:
> I think the one called from `bt_mesh_start` is now covered by your change in `prov_link_opened` callback.

@PavelVPV I'm not sure if there's any difference in the terms of the pb_gatt_srv status once provisioning is completed (i. e. over PB-ADV). This is due to the `link_accept` cb being called for all the suspended bearers in `prov_link_closed`. Maybe we should either check if the device has been provisioned here, or add an early return to the `link_accept` cbs? (Not fully sure how this works for the RPR case, but I would guess we want to execute the cb for this anyway?)

PavelVPV commented on subsys/bluetooth/mesh/provisionee.c:
```diff
@@ -657,6 +671,22 @@ static void prov_link_closed(enum prov_bearer_link_status status)
 	}
 
 	bt_mesh_prov_reset_state();
+
```
Shouldn't we exit earlier here if provisioning was successful?

HaavardRei replied:
Agreed, but I think we should re-enable RPR first (if we cancelled it) since this still has some available functionality (reprovisioning)?

PavelVPV replied:
> I think we should re-enable RPR first (if we cancelled it) since this still has some available functionality (reprovisioning)?

True

HaavardRei commented:
AFAIS (and from local testing) the call in settings.c would always result in `bt_mesh_pb_gatt_srv_disable` returning early due to the service not yet being registered, so I removed it!

HaavardRei commented on subsys/bluetooth/mesh/prov_bearer.h:
```diff
@@ -76,6 +76,12 @@ struct prov_bearer {
 
+	/** @brief Disable link establishment as a provisionee.
+	 *
+	 *  Stops accepting link open messages and hides from provisioners.
```
The provisionee stops sending unprovisioned beacons (PB-ADV) or advs (PB-GATT).

HaavardRei commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -634,15 +634,10 @@ int bt_mesh_start(void)
-		}
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY) &&
+	    (!IS_ENABLED(CONFIG_BT_MESH_PROV) || !bt_mesh_prov_active() ||
+	     bt_mesh_prov_link.bearer->type == BT_MESH_PROV_ADV)) {
```
Added `bt_mesh_prov_link.bearer->type == BT_MESH_PROV_REMOTE` to the if.

HaavardRei commented:
I needed to make some changes to the cancelling logic for RPR. Since pb-remote can use both the adv and gatt bearers, I changed it so a pb-gatt link disables pb-adv and vice versa, whereas for pb-remote both are kept in their current state.

HaavardRei commented:
> The `bt_mesh_pb_gatt_srv_disable` called from [bt_mesh_start](https://github.com/zephyrproject-rtos/zephyr/blob/4bb5ffd7867d9b497fa8f03d73973ab2e32c69ba/subsys/bluetooth/mesh/main.c#L640) was meant to disable active PB-GATT if device was provisioned over PB-ADV. Since now PB-GATT [is stopped](https://github.com/zephyrproject-rtos/zephyr/blob/d9539319e789ec79e9531bd00c5c807be6189e76/subsys/bluetooth/mesh/provisionee.c#L700-L703) if provisioning is started over PB-ADV, when provisioning completes, PB-GATT stays disabled. Thus, the `bt_mesh_pb_gatt_srv_disable` from `bt_mesh_start` is not needed anymore.

For some reason, removing this call breaks the gatt suspend/resume and the advertiser/proxy_mixin test. I reverted the change in main.c, and will create a followup task to investigate this.

HaavardRei commented:
DNM until https://github.com/zephyrproject-rtos/zephyr/pull/87428 is merged and PR is rebased, but should be fine to review.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -290,7 +302,21 @@ static void unprovisioned_beacon(uint8_t uuid[16],
 	}
-	bt_mesh_provision_adv(uuid, 0, prov_addr, 0);
+
+	if (memcmp(uuid, prov_uuid, 16) == 0) {
+		ASSERT_FALSE_MSG((prov_started && (k_uptime_delta(&prov_started) > 100)),
+				 "Received PB-ADV beacon after provisioning started.");
+	}
```
Move this to a generic function which then can be used for both, pb-gatt and pb-adv. Move `100` to another macro and add a comment that it is taken taken randomly but reasonably small. (We can calculate it though here, not sure if we need this now. But if test starts failing, we can do this).

PavelVPV reviewed (changes requested):
Sorry, I missed that unprovisioned beacon interval is quite long. To test that no beacons are actually sent, I suggest to start provisioning with `input_actions` set to e.g. `BT_MESH_ENTER_NUMBER`  (reference code: https://github.com/zephyrproject-rtos/zephyr/blob/2338f8efc5b29b5fb0fde1d9d8cf0a0046b8d9ea/tests/bsim/bluetooth/mesh/src/test_provision.c#L461-L466, set other parameters from this code to 0) and schedule input with delay equal to `K_SECONDS(CONFIG_BT_MESH_UNPROV_BEACON_INT + 1)` here: https://github.com/zephyrproject-rtos/zephyr/blob/2338f8efc5b29b5fb0fde1d9d8cf0a0046b8d9ea/tests/bsim/bluetooth/mesh/src/test_provision.c#L357

Rest comments are minor.

HaavardRei committed:
Bluetooth: Mesh: Disable prov bearers on link

Previously, PB-GATT unprovisioned advs would continue being sent if a
device is provisioned or in the process of being provisioned (an active
provisioning link) over PB-ADV.

This commit introduces the `link_cancel` callback for the `prov_bearer`
struct. When a provisioning link is established over a prov bearer, all
other (active) prov bearers will be suspended through this callback, and
re-enabled when the link is closed. An exception is made for PB-REMOTE,
which can utilize the other bearers.

The call to `bt_mesh_pb_gatt_srv_disable` in settings.c is removed. This
is called in `mesh_commit`, and is not needed as this is only used
during boot, at which point the Mesh provisioning service has not yet
been enabled/registered, meaning the function does nothing.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Mesh: Fix typo in prov bsim

Changes Brearer --> Bearer.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Mesh: Test pb cancelling in bsim

Adds a provision test to check that provisioning bearers are cancelled
when a provisioning link is opened. Modifies the functionality for
receiving unprovisioned beacons to fail the test if a unprovisioned
beacon is received from the current provisionee after a threshold.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented on tests/bsim/bluetooth/mesh/tests_scripts/provision/prov_multi.sh:
```diff
@@ -12,20 +12,30 @@ RunTest mesh_prov_pb_adv_multi \
+	prov_device_no_oob \
+	prov_device_no_oob \
+	prov_device_no_oob \
+	-- -argstest prov-bearer=3 prov-to-use=2
```
Done!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -290,7 +302,21 @@ static void unprovisioned_beacon(uint8_t uuid[16],
+			prov_started = k_uptime_get();
+			memcpy(prov_uuid, uuid, 16);
+		}
+	}
```
I changed it so that `provision` returns void. This is since the function would return -EBUSY when receiving successive beacons, and also when running the `prov_oob_auth(_ignore)_oob_pk.sh ` tests locally it would return `-EINVAL` after passing the test, causing a failure... (probably due to some reset functionality).
We could add error handling if you think this is needed (it wasn't present before), but I would suggest doing so in a separate task.

alxelax commented on subsys/bluetooth/mesh/pb_gatt.c:
```diff
@@ -246,6 +251,7 @@ const struct prov_bearer bt_mesh_pb_gatt = {
 #endif
 #if defined(CONFIG_BT_MESH_PB_GATT)
```
I guess this is useless conditional compilation since CONFIG_BT_MESH_PB_GATT is always `y` for this file, but it is out of scope of your changes.

alxelax commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -104,10 +104,6 @@ static int mesh_commit(void)
 		return 0;
 	}
 
-	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
```
Why did you remove this? Could you elaborate a bit?

HaavardRei replied:
I added this explanation in the commit message (Realized I'm missing a "call", will fix):
> The [call] to `bt_mesh_pb_gatt_srv_disable` in settings.c is removed. This is
called in `mesh_commit`, and is not needed as this is only used during
boot, at which point the Mesh provisioning service has not yet been
enabled/registered, meaning the function does nothing.

Is this sufficient?

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/provision/pb_cancel.sh:
```diff
@@ -0,0 +1,27 @@
+# Test procedure:
+# 1. The provisioner self-provisions and starts scanning for unprovisioned devices.
+# 2. The provisionee enables both the PB-ADV and PB-GATT bearers.
+# 3. The provisioner finds the unprovisioned device and provisions it with OOB authentication.
```
nice trick :)

HaavardRei replied:
Twas @PavelVPV 's suggestion :)

alxelax reviewed (commented):
Please fix comment in prov_bearer.h. The others look Ok to me.

HaavardRei commented on subsys/bluetooth/mesh/prov_bearer.h:
```diff
@@ -76,6 +76,12 @@ struct prov_bearer {
 
+	/** @brief Disable link establishment as a provisionee.
+	 *
+	 *  Stops accepting link open messages and hides from provisioners.
```
True, changed it!
